src/transformers/modeling_reformer.py

```diff
@@ -0,0 +1,1757 @@
+        return config.local_attn_chunk_length
+    elif len(set(attn_types)) == 2 and set(attn_types) == set(["lsh", "local"]):
+        return np.lcm(config.lsh_attn_chunk_length, config.local_attn_chunk_length)
+    else:
+        raise NotImplementedError(
+            "Only attn layer types 'lsh' and 'local' exist, but `config.attn_layers`: {}. Select attn layer types from ['lsh', 'local'] only.".format(
+                config.attn_layers
+            )
+        )
+
+
+class AxialPositionEmbeddings(nn.Module):
+    """Constructs axial position embeddings. Useful for very long input
+    sequences to stay memory and computational efficient
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        self.axial_pos_shape = config.axial_pos_shape
+        self.axial_pos_embds_dim = config.axial_pos_embds_dim
+        self.dropout = config.hidden_dropout_prob
+
+        self.least_common_mult_chunk_length = _get_least_common_mult_chunk_len(config)
+        self.weights = nn.ParameterList()
+
+        assert (
+            sum(self.axial_pos_embds_dim) == config.hidden_size
+        ), "Make sure that config.axial_pos_embds factors: {} sum to config.hidden_size: {}".format(
+            self.axial_pos_embds_dim, config.hidden_size
+        )
+
+        # create weights
+        for axis, axial_pos_embd_dim in enumerate(self.axial_pos_embds_dim):
+            # create shape
+            ax_shape = [1] * len(self.axial_pos_shape)
+            ax_shape[axis] = self.axial_pos_shape[axis]
+            ax_shape = tuple(ax_shape) + (axial_pos_embd_dim,)
+
+            # create tenser and init
+            self.weights.append(nn.Parameter(torch.ones(ax_shape, dtype=torch.float32)))
+
+    def forward(self, position_ids):
+        # broadcast weights to correct shape
+        batch_size = position_ids.shape[0]
+        sequence_length = position_ids.shape[1]
+
+        broadcasted_weights = [
+            weight.expand((batch_size,) + self.axial_pos_shape + weight.shape[-1:]) for weight in self.weights
+        ]
+
+        if self.training is True:
+            assert (
+                reduce(mul, self.axial_pos_shape) == sequence_length
+            ), "Make sure that config.axial_pos_shape factors: {} multiply to sequence length: {}".format(
+                self.axial_pos_shape, sequence_length
+            )
+            if self.dropout > 0:
+                weights = torch.cat(broadcasted_weights, dim=-1)
+                # permute weights so that 2D correctly drops dims 1 and 2
+                perm_weigthts = weights.permute(0, 3, 2, 1)
+                # drop entire matrix of last two dims (prev dims 1 and 2)
+                drop_perm_weights = nn.functional.dropout2d(perm_weigthts, self.dropout, training=self.training)
+                drop_weights = drop_perm_weights.permute(0, 3, 2, 1)
+
+                position_encodings = torch.reshape(drop_weights, (batch_size, sequence_length, -1))
```

Comment:
is this equivalent to `.view`?

Reply:
`reshape` automatically applies the `contiguous()` function if necessary whereas `view` does not but would throw an error